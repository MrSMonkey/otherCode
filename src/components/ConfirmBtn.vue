<!--
 * @Description: 底部确认按钮
 * @Author: chenmo
 * @LastEditors: chenmo
 * @Date: 2019-02-28 13:56:48
 * @LastEditTime: 2019-03-04 16:15:02
 -->

<template>
  <section class="plot-footer">
    <van-button size="normal" type="default" @click="plotCancel">{{cancelText}}</van-button>
    <van-button size="normal" type="default"  @click="confirm" :loading="loading" v-if="isActiveKey" :loading-text="loadingText" class="plot-active">
      <slot name="confirm"></slot>
    </van-button>
    <van-button size="normal" type="default" :url="url" v-else><slot name="confirm"></slot></van-button>
  </section>
</template>

<script lang="ts">
import { Component, Vue, Prop, Emit, Watch } from 'vue-property-decorator';

@Component({
  name: 'ConfirmBtn'
})
export default class ConfirmBtn extends Vue {
  @Prop({type: Boolean, default: false})
  private loading: boolean;

  @Prop({type: Boolean, default: false})
  private isActive: boolean;

  @Prop({ type: String, default: '' })
  private url: string;

  @Prop({ type: String, default: '加载中' })
  private loadingText: string;

  @Prop({ type: String, default: '取消' })
  private cancelText: string;

  @Prop({ type: String, default: '确认' })
  private confirmText: string;

  private isActiveKey: boolean = false;

  private mounted() {
    this.isActiveKey = this.isActive;
  }

  @Emit('confirm')
  private confirm() {
    return;
  }

  @Emit('plotCancel')
  private plotCancel() {
    return;
  }

  @Watch('isActive')
  private handlerisActive(newVal: boolean) {
    this.isActiveKey = newVal;
  }
}
</script>

<style lang="stylus" rel="stylesheet/stylus">
@import '../assets/stylus/main.styl'
.plot-footer
  position absolute
  bottom 0
  left 0
  display -webkit-flex
  display flex
  justify-content space-between
  width 100%
  height vw(46)
  align-items center
  z-index 1000
  button
    display inline-block
    width 100%
    font-size 14px
    text-align center
    height 100%
    border-radius 0
    &:nth-child(1)
      background #fff
      color $main-color
      border 1px solid #fff
    &:nth-child(2)
      background $disabled-color
      color #fff
      border 1px solid $disabled-color
  .plot-active
    background $main-color !important
    border 1px solid $main-color !important
</style>