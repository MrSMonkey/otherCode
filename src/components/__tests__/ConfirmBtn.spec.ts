/*
 * @Description: 底部确认按钮单元组件测试
 * @Author: chenmo
 * @Date: 2019-03-26 15:28:22
 * @Last Modified by: chenmo
 * @Last Modified time: 2019-03-27 11:51:02
 */

import Vue from 'vue';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import ConfirmBtn from '@/components/ConfirmBtn.vue';

// 独立构建Vue构造函数
const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(VueRouter);

describe('对ConfirmBtn组件进行的单元测试', () => {
  let store: any;
  let wrapper: any;
  let propsData: any;

  beforeEach(() => {
    propsData = {
      loadingText: '加载中',
      cancelText: '取消',
      confirmText: '确认',
      isActiveKey: false,
      loading: false,
      url: '/home'
    };

    store = new Vuex.Store({
      state: {
        global: {}
      }
    });

    // 挂载组件
    wrapper = shallowMount(ConfirmBtn, {
      store,
      propsData,
      localVue
    });
  });

  afterEach(() => {
    // 销毁实例
    wrapper.destroy();
  });

  test('ConfirmBtn', () => {
    // 测试实例是Vue组件
    expect(wrapper.isVueInstance()).toBeTruthy();
    // 存在 BottomBtn 组件
    expect(wrapper.is(ConfirmBtn)).toBeTruthy();
    // 验证组件是否渲染
    expect(wrapper.contains('.plot-footer')).toBeTruthy();
  });

  test('接收到了confirmText作为Props', () => {
    expect(wrapper.props().confirmText).toContain('确认');
  });

  // 生成测试快照
  test('create expected html structure', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
