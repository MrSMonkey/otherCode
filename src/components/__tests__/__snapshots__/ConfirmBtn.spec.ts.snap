// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`对ConfirmBtn组件进行的单元测试 create expected html structure 1`] = `
<section
  class="plot-footer"
>
  <van-button
    size="normal"
    type="default"
  >
    取消
  </van-button>
   
  <van-button
    size="normal"
    type="default"
    url="/home"
  />
</section>
`;
